### PR TITLE
react-widgets: Improves DateTimePicker parse prop type

### PR DESCRIPTION
**Pull request by wavebeem**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: (https://github.com/jquense/react-widgets/blob/master/packages/react-widgets/src/DateTimePicker.js#L604)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

## Comments

It's valid for the `parse` prop on a react-widgets/DateTimePicker to return undefined (technically any falsy value) in order to disregard the text input value (i.e. parse failure). Like if someone just types "foo" into a DateTimePicker, then the input should be cleared upon blurring the field. But the existing DT types don't account for that return type.

I ran prettier as the documentation suggested, which updated most of the files within types/react-widgets. I assume this is OK, but let me know if there's an issue.

I have left comments on the two non-prettier changes I'm actually submitting.

### CODE REVIEW

File: types/react-widgets/lib/DateTimePicker.d.ts [hunk 1]
```diff
@@ -88,7 +88,7 @@ declare namespace DateTimePicker {
          * parsing yourself. When parse is unspecified and the format prop is a String parse will
          * automatically use that format as its default
          */
-        parse?: ((str: string) => Date) | string[] | string;
+        parse?: ((str: string) => Date | undefined) | string[] | string;
```
Comment (wavebeem) [inline types/react-widgets/lib/DateTimePicker.d.ts:91]: This is the real bug fix: the parse callback should be able to return `undefined`

File: types/react-widgets/react-widgets-tests.tsx [hunk 2]
```diff
@@ -147,6 +103,16 @@ class Test extends React.Component<React.Props<{}>> {
...
+                <div>
+                    <DateTimePicker
+                        parse={(str: string): Date | undefined => {
+                            if (str) {
+                                return new Date(str);
+                            }
+                            return undefined;
+                        }}
+                    />
+                </div>
```
Comment (wavebeem) [inline types/react-widgets/react-widgets-tests.tsx:115]: And this is the test for the updated type.

File: types/react-widgets/index.d.ts [hunk 3]
```diff
@@ -14,10 +14,10 @@
...
 // TypeScript Version: 2.8
 
-export import Calendar = require('./lib/Calendar');
-export import Combobox = require('./lib/Combobox');
-export import DateTimePicker = require('./lib/DateTimePicker');
-export import DropdownList = require('./lib/DropdownList');
-export import Multiselect = require('./lib/Multiselect');
-export import NumberPicker = require('./lib/NumberPicker');
-export import SelectList = require('./lib/SelectList');
+export import Calendar = require("./lib/Calendar");
```
Comment (r3nya) [inline types/react-widgets/index.d.ts:17]: What the point to commit these changes for codestyle? :)

**Review by r3nya — APPROVED** [review]

File: types/react-widgets/index.d.ts (hunk 3 above)
Comment (wavebeem) [reply types/react-widgets/index.d.ts:17]: The README suggested I run Prettier on the package folder, so I committed all the changes. https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes

**Comment by wavebeem** [comment]
Ready to merge

**Comment by wavebeem** [comment]
Thanks @r3nya for reviewing my PR!

**Review by sanyatuning — APPROVED** [review]
Lgtm

File: types/react-widgets/index.d.ts (hunk 3 above)
Comment (r3nya) [reply types/react-widgets/index.d.ts:17]: Gotcha! Thanks for your answer :)

**Pull request merged**